### PR TITLE
[PSTL] Use env based overload for `DeviceFindIf`

### PR DESCRIPTION
**Pull request by miscco**
We want to use env based API to ensure that we take advantage of user provided tunings

### CODE REVIEW

**Review by gonidelis — CHANGES_REQUESTED** [review]
why are we substituting the stream overloads with env overloads? don't we already have existing env overloads at home?

File: cub/cub/device/device_find.cuh [hunk 1]
```diff
@@ -107,26 +110,39 @@ struct DeviceFind
...
+            typename EnvT = ::cuda::std::execution::env<>>
   CUB_RUNTIME_FUNCTION static cudaError_t FindIf(
     void* d_temp_storage,
     size_t& temp_storage_bytes,
     InputIteratorT d_in,
     OutputIteratorT d_out,
     ScanOpT scan_op,
     NumItemsT num_items,
-    cudaStream_t stream = nullptr)
+    const EnvT& env = {})
```
Comment (bernhardmgruber) [inline cub/cub/device/device_find.cuh:130]: Remark: Changing to an environment here would require several new tests to be added, not only about whether various stream types can be passed here, but also whether a tuning passed here would be applied. If we proceed with more such changes across the entire CUB device layer, we would need to write a TON of redundant tests.

I think we should try to move as much logic as possible into the `dispatch` function, so the device layer becomes so thin that we can reasonably only cover one of the overloads (e.g. the one without temp storage args) with all tests.

**Comment by miscco** [comment]
> why are we substituting the stream overloads with env overloads? don't we already have existing env overloads at home?

We have only those that take care of all the memory management. But we also need the env for those where the user provides memory

File: cub/cub/device/device_find.cuh (hunk 1 above)
Comment (miscco) [reply cub/cub/device/device_find.cuh:130]: The way the refactoring is structured is to just forward to the dispatch layer.

The forward is exactly the same as with the current env overloads, so I am not sure whether we would need any more testing.

My suggestion is be to split the effort in two. There is no reason why we need to block refactoring the top level APIs on the dispatch refactoring, given we only dispatch to the same function anyway.

File: cub/cub/detail/env_dispatch.cuh [hunk 2]
```diff
@@ -37,7 +37,7 @@ namespace detail
 //! @param env The execution environment
 //! @param algorithm_callable Callable that invokes the algorithm implementation with determinism specified
 template <typename EnvT, typename AlgorithmCallable>
-CUB_RUNTIME_FUNCTION static cudaError_t dispatch_with_env(EnvT env, AlgorithmCallable&& algorithm_callable)
+CUB_RUNTIME_FUNCTION static cudaError_t dispatch_with_env(const EnvT& env, AlgorithmCallable&& algorithm_callable)
```
Comment (gonidelis) [inline cub/cub/detail/env_dispatch.cuh:40]: super nice - lets adopt

File: cub/cub/detail/env_dispatch.cuh [hunk 3]
```diff
@@ -86,6 +87,48 @@ CUB_RUNTIME_FUNCTION static cudaError_t dispatch_with_env_and_tuning(EnvT env, A
...
+//! Generic environment-based algorithm dispatch wrapper
+//!
+//! Handles common boilerplate for env-based algorithms with user provided memory:
+//! - Query stream, and tuning from environment
+//! - Single-phase call passing user provided memory and size
+//!
+//! @param env The execution environment
+//! @param[in] d_temp_storage @devicestorage
+//! @param[in,out] temp_storage_bytes Reference to size in bytes of `d_temp_storage` allocation
+//! @param algorithm_callable Callable that invokes the algorithm implementation with determinism specified
```
Comment (gonidelis) [inline cub/cub/detail/env_dispatch.cuh:103]: minor: not sure why do we add docs in the dispatch_with_env helpers in the first place

File: cub/cub/detail/env_dispatch.cuh [hunk 4]
```diff
@@ -86,6 +87,48 @@ CUB_RUNTIME_FUNCTION static cudaError_t dispatch_with_env_and_tuning(EnvT env, A
...
+
+template <typename DefaultPolicySelector, typename EnvT, typename AlgorithmCallable>
+CUB_RUNTIME_FUNCTION static cudaError_t dispatch_with_env_and_tuning(
+  void* d_temp_storage, size_t& temp_storage_bytes, const EnvT& env, AlgorithmCallable&& algorithm_callable)
+{
+  return detail::dispatch_with_env(
+    d_temp_storage,
+    temp_storage_bytes,
+    env,
+    [&]([[maybe_unused]] auto tuning_env, void* d_temp_storage, size_t& temp_storage_bytes, cudaStream_t stream) {
```
Comment (gonidelis) [inline cub/cub/detail/env_dispatch.cuh]: nit: isn't that lambda shadowing because of `[&]`? same args are being used on the signature

Comment (gonidelis) [reply cub/cub/detail/env_dispatch.cuh]: i'd do sth like:

```suggestion
[&algorithm_callable]([[maybe_unused]] auto tuning_env, void* d_temp_storage, size_t& temp_storage_bytes, cudaStream_t stream) {
```
to make it clearer for future maintainers

File: cub/cub/device/device_find.cuh [hunk 5]
```diff
@@ -223,27 +244,33 @@ struct DeviceFind
...
 
     using RangeOffsetT  = detail::choose_offset_t<RangeNumItemsT>;
     using ValuesOffsetT = detail::choose_offset_t<ValuesNumItemsT>;
 
-    if (d_temp_storage == nullptr)
-    {
-      temp_storage_bytes = 1;
-      return cudaSuccess;
-    }
+    return detail::dispatch_with_env(
```
Comment (gonidelis) [inline cub/cub/device/device_find.cuh:254]: again thinking for future developers that will jump on CUB ship, our high level bodies start to look really ugly, no complain for now. just laying this one out in the field

File: cub/test/catch2_test_device_find_env.cu [hunk 6]
```diff
@@ -169,6 +285,99 @@ C2H_TEST("Device LowerBound uses environment", "[find][device]")
   REQUIRE(d_output == expected);
 }
 
+C2H_TEST("Device LowerBound works with user provided memory and environment", "[find][device]")
+{
+  auto d_range                     = c2h::device_vector<int>{0, 2, 4, 6, 8};
+  auto d_values                    = c2h::device_vector<int>{1, 3, 5, 7};
```
Comment (gonidelis) [inline cub/test/catch2_test_device_find_env.cu:295]: nit: cover edge cases too

```suggestion
  auto d_values = c2h::device_vector<int>{1, 2, 3, 5};
```

same for upper_bound

File: cub/test/catch2_test_device_find_env.cu [hunk 7]
```diff
@@ -169,6 +285,99 @@ C2H_TEST("Device LowerBound uses environment", "[find][device]")
...
+  auto d_output                    = c2h::device_vector<int>(4);
+  c2h::device_vector<int> expected = {1, 2, 3, 4};
```
Comment (gonidelis) [inline cub/test/catch2_test_device_find_env.cu:297]: if you decide to accept the above this should become

`c2h::device_vector<int> expected = {1, 1, 2, 3};`

**Review by gonidelis — APPROVED** [review]

File: cub/test/catch2_test_device_find_env.cu (hunk 6 above)
Comment (miscco) [reply cub/test/catch2_test_device_find_env.cu:295]: This is an exact copy of the iterators test case. I believe we should spen some time to extrack Lower / Upper bounds testing into a separate file and test them more thoroughly. 

I do not want to add new test patterns here though

File: cub/cub/device/device_find.cuh (hunk 5 above)
Comment (miscco) [reply cub/cub/device/device_find.cuh:254]: sad but true

File: cub/cub/detail/env_dispatch.cuh (hunk 3 above)
Comment (miscco) [reply cub/cub/detail/env_dispatch.cuh:103]: I copied it out of habbit, does not hurt because its not public

File: cub/cub/detail/env_dispatch.cuh (hunk 4 above)
Comment (miscco) [reply cub/cub/detail/env_dispatch.cuh]: yeah you are right, good catch

**Review by davebayer — APPROVED** [review]

File: cub/cub/device/device_find.cuh (hunk 1 above)
Comment (bernhardmgruber) [reply cub/cub/device/device_find.cuh:130]: I will keep your reply in mind and send any resulting bugs your way ;)
